homeassistant/components/rainbird/calendar.py

```diff
@@ -0,0 +1,159 @@
+"""Rain Bird irrigation calendar."""
+
+from __future__ import annotations
+
+from datetime import datetime, timedelta
+import logging
+
+from homeassistant.components.calendar import CalendarEntity, CalendarEvent
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+from homeassistant.util import dt as dt_util
+
+from .const import DOMAIN
+from .coordinator import RainbirdScheduleUpdateCoordinator
+
+_LOGGER = logging.getLogger(__name__)
+
+# Polling the calendar requires a number of RPC calls to fetch the data and
+# we don't expect the calendar to change often, so use a long schedule.
+MIN_TIME_BETWEEN_UPDATES = timedelta(minutes=30)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up entry for a Rain Bird irrigation calendar."""
+    coordinator = hass.data[DOMAIN][config_entry.entry_id].coordinator
+    model = await coordinator.controller.get_model_and_version()
+    if not model.model_info.max_programs:
+        return
+
+    schedule_coordinator = hass.data[DOMAIN][config_entry.entry_id].schedule_coordinator
+    async_add_entities(
+        [
+            RainBirdCalendarEntity(
+                schedule_coordinator,
+                coordinator.serial_number,
+                coordinator.device_info,
+            )
+        ]
+    )
+
+
+class RainBirdCalendarEntity(
+    CoordinatorEntity[RainbirdScheduleUpdateCoordinator], CalendarEntity
+):
+    """A calendar event entity."""
+
+    _attr_has_entity_name = True
+    _attr_icon = "mdi:sprinkler"
+
+    def __init__(
+        self,
+        coordinator: RainbirdScheduleUpdateCoordinator,
+        serial_number: str,
+        device_info: DeviceInfo,
+    ) -> None:
+        """Create the Calendar event device."""
+        super().__init__(coordinator)
+        self._event: CalendarEvent | None = None
+        self._attr_unique_id = serial_number
+        self._attr_device_info = device_info
+
+    @property
+    def should_poll(self) -> bool:
+        """Enable polling for the entity.
+
+        The coordinator has its own polling interval for querying the calendar,
+        and this entity updates its own state more frequently (e.g. to update when
+        an event starts).
+        """
+        return True
```

Comment:
I'm a bit confused how this works considering the update method is overridden to do nothing below?

Reply:
Here is some context:
https://github.com/home-assistant/core/pull/92712

I remember now we were holding on doing more with this pattern and coming up with a solution in the calendar integration.

I suspect the solution looks something like having calendar integration automatically update state based on the upcoming calendar event time. If that sounds reasonable to you, I can explore that and get Frenck's ACK also, then come back to this.

Reply:
It would be better if it only write state based on an event/predicted change/etc instead of blindly doing it at a 1m interval.

Reply:
I'm going to propose a change to calendar to handle this and come back to this after.

Reply:
Calendar entity was simplified in #98037 to handle these state transitions, so this was all streamlined.